ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/tracing/vm/VmTraceGenerator.java

```diff
@@ -0,0 +1,284 @@
+    final VmOperation op = buildVmOperation();
+    final VmOperationExecutionReport report = generateExecutionReport();
+    generateTracingMemory(report);
+    generateTracingPush(report);
+    generateTracingStorage(report);
+    handleDepthIncreased(op, report);
+    handleDepthDecreased();
+    completeStep(op, report);
+  }
+
+  private boolean mustIgnore(final TraceFrame frame) {
+    return "STOP".equals(frame.getOpcode()) && transactionTrace.getTraceFrames().size() == 1;
+  }
+
+  private void completeStep(final VmOperation op, final VmOperationExecutionReport report) {
+    // add the operation representation to the list of traces
+    op.setVmOperationExecutionReport(report);
+    currentTrace.add(op);
+    currentIndex++;
+    lastCapturedStorage = currentTraceFrame.getStorage();
+  }
+
+  private void handleDepthIncreased(final VmOperation op, final VmOperationExecutionReport report) {
+    // check if next frame depth has increased i.e the current operation is a call
+    if (maybeNextFrame.map(next -> next.isDeeperThan(currentTraceFrame)).orElse(false)) {
+      maybeNextFrame.ifPresent(
+          nextFrame -> op.setCost(nextFrame.getGasRemaining().toLong() + op.getCost()));
+      final VmTrace newSubTrace = new VmTrace();
+      parentTraces.addLast(newSubTrace);
+      findLastFrameInCall(currentTraceFrame, currentIndex)
+          .ifPresent(
+              lastFrameInCall -> {
+                report.setUsed(lastFrameInCall.getGasRemaining().toLong());
+                lastFrameInCall
+                    .getStack()
+                    .filter(stack -> stack.length > 0)
+                    .map(stack -> stack[stack.length - 1])
+                    .map(last -> Quantity.create(UInt256.fromHexString(last.getHexString())))
+                    .ifPresent(report::addPush);
+              });
+      op.setSub(newSubTrace);
+    }
+  }
+
+  private void handleDepthDecreased() {
+    // check if next frame depth has decreased i.e the current operation closes the parent trace
+    if (maybeNextFrame.map(next -> next.isLessDeepThan(currentTraceFrame)).orElse(false)) {
+      currentTrace = parentTraces.removeLast();
+    }
+  }
+
+  private VmOperation buildVmOperation() {
+    final VmOperation op = new VmOperation();
+    // set gas cost and program counter
+    op.setCost(currentTraceFrame.getGasCost().orElse(Gas.ZERO).toLong());
+    op.setPc(currentTraceFrame.getPc());
+    return op;
+  }
+
+  private VmOperationExecutionReport generateExecutionReport() {
+    final VmOperationExecutionReport report = new VmOperationExecutionReport();
+    // set gas remaining
+    report.setUsed(
+        currentTraceFrame.getGasRemaining().toLong()
+            - currentTraceFrame.getGasCost().orElse(Gas.ZERO).toLong());
+    return report;
+  }
+
+  private void generateTracingMemory(final VmOperationExecutionReport report) {
+    if (currentTraceFrame.getDepth() == maybeNextFrame.map(TraceFrame::getDepth).orElse(0)) {
```

Comment:
Is there a good rationale for this conditional?  Looks like this means any kind of CALL-like operation won't display any memory changes?  What happens if the last operation before a call returns writes to memory?  It looks like we wouldn't include those changes.  Can we add a test for this?

Reply:
my bad it is a mistake. with the previous approach of retrieving changed memory it was required but not anymore. I have removed this condition.